runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java

```diff
@@ -284,8 +286,16 @@ public void flatMap(T s, Collector<WindowedValue<T>> collector) throws Exception
               }
             }).assignTimestampsAndWatermarks(new IngestionTimeExtractor<WindowedValue<T>>());
       } else {
-        source = context.getExecutionEnvironment()
-            .addSource(new UnboundedSourceWrapper<>(context.getPipelineOptions(), transform));
+        try {
+          UnboundedSourceWrapper<T, UnboundedSource.CheckpointMark> sourceWrapper =
+              new UnboundedSourceWrapper<>(
+                  context.getPipelineOptions(),
+                  (UnboundedSource) transform.getSource(),
+                  context.getExecutionEnvironment().getParallelism());
```

Comment:
Some of this may not translate directly because of the differences between Dataflow service and Flink clusters.

* We do indeed perform initial splitting of `UnboundedSource` at job submission time, aka when the user issues a job before submitting it to the service. This does not generally make us happy. As a signal for how far to split, we use the [`DataflowPipelineOptions.getMaxNumWorkers()`](https://github.com/apache/incubator-beam/blob/e3105c8e109535f801fd145b91b0c7aa93b86d1a/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/CustomSources.java#L68) setting.

These splits are used for the entire lifetime of the job. (Hence the extra `*3` over max workers, so that we can rebalance hot splits, etc.)

We would like to additionally add dynamic splitting, such that a hot split could be subsplit. The APIs for this are not quite designed -- somehow you'd have to split given an `UnboundedSource` and a `CheckpointMark` and get 2 new pairs of the same.

We have not been planning to upgrade these APIs with these features, although we might if it becomes important to the community. What we'd like to do is a holistic re-evaluation of how sources (and not just sources -- also intermediate DoFns that have a high-fan-out of data!) can be checkpointed, restarted, and split off. See [BEAM-65](https://issues.apache.org/jira/browse/BEAM-65)